### PR TITLE
[WGSL] Add overload resolution for (non-parameterized) value constructors

### PR DESCRIPTION
**Pull request by tadeuzagallo**
#### 47c9e4a9f264797e16760164e913cf7c6aef8abd
<pre>
[WGSL] Add overload resolution for (non-parameterized) value constructors
<a href="https://bugs.webkit.org/show_bug.cgi?id=254419">https://bugs.webkit.org/show_bug.cgi?id=254419</a>
rdar://107185949

Reviewed by Myles C. Maxfield.

This will be used to add support to e.g. vec4(...)

* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::visit):

Canonical link: <a href="https://commits.webkit.org/262199@main">https://commits.webkit.org/262199@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d4084e38259574f1424fea16840d01bff6991152

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/488 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/503 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/526 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/526 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/437 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/549 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/590 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/684 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/474 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/543 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/518 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/463 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/510 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/432 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/480 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/481 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/427 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/474 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/206 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/470 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebGPU/WGSL/TypeCheck.cpp [hunk 1]
```diff
@@ -318,9 +318,35 @@ void TypeChecker::visit(AST::IdentifierExpression& identifier)
...
+        arguments.append(infer(argument));
+
+    auto& target = call.target();
+    if (is<AST::NamedTypeName>(target)) {
+        auto& namedTarget = downcast<AST::NamedTypeName>(target);
+        auto* result = chooseOverload(namedTarget.name(), arguments);
+        if (result)
+            inferred(result);
+        else {
+            StringPrintStream argumentsString;
```
Comment (litherum) [inline Source/WebGPU/WGSL/TypeCheck.cpp:340]: Not StringBuilder?

I'm consistently bewildered at the different in coding conventions between JSC and WebCore

**Review by litherum — APPROVED** [review]
No test?

**Review by djg — APPROVED** [review]
:shipit:

File: Source/WebGPU/WGSL/TypeCheck.cpp (hunk 1 above)
Comment (tadeuzagallo) [reply Source/WebGPU/WGSL/TypeCheck.cpp:340]: We seem to have both in WebGPU, I just opened a random file and used the same things that was being used 🤷‍♂️

**Comment by tadeuzagallo** [comment]
> No test?

Indeed, I need to start adding more tests. That's part of why I was working on the standalone `wgslc`, since that'd make it trivial to test this, but I'm sure there are other ways we can test it. I'll look into it before landing.

**Comment by tadeuzagallo** [comment]
I'll land this as is, since there's no overload declarations to test yet, but I'll add tests in the PR that introduces the type declarations

**Pull request merged**